### PR TITLE
docs: Clarify important points about parametric types

### LINKED ISSUE #43811: Better introduction to parametric types in the manual
After a discussion in IRC today I realised that there were a few things about parametric types that aren't easily understood. It seems that the manual could help a bit more in introducing parametric types. I'm opening this issue to outline some of the things I think could be documented better, and also to gather ideas or examples that might help to write a more hands-on parametric types section in the manual. Not sure if this belongs on the forum but I am thinking of linking a PR at some point.

For the parametric type `Foo` and an instance `foo`, I think the docs could clarify
- why both `isabstracttype(Foo)` and `isconcretetype(Foo)` return false for the parametric type `Foo`, i.e. that parametric types are neither abstract nor concrete
- the difference between the following definitions of `Foo`

```julia
struct Foo{T}
    a::T
    b::T
end

struct Foo{T1, T2}
    a::T1
    b::T2
end
```
- why `typeof(foo) <: Foo` is true, however `supertype(typeof(foo)) == Foo` is false (it's `Any`)
- why `Foo isa DataType` is false

and (more visibly) inform
- that parametric types cannot be subtyped
- that reified instances of parametric types are of a concrete type, therefore `typeof(foo) != Foo` where `foo` is an instance of `Foo`
- that `bar isa Foo` must be used to test if `foo` is an instance of `Foo`

I also think there should be one example on how to set up a simple but complete type hierarchy (without going into all of the gritty details of the Advanced Types section). Something like the comment by wasshin in #4935.

There might be some other things I've forgotten, feel free to suggest related improvements. I feel that parametric types are one of the major features of Julia's type system, however they can be confusing when coming from languages with more traditional/simple OOP systems. Understanding parametric types should help to reduce common antipatterns like over-constraining argument types, and help people to leverage the dispatch system (what are all those `<:` needed for anyway?)

### PR DESCRIPTION
**Pull request by adigitoleo**
Closes #43811.

- The intro section felt a bit long-winded, I've made some changes there first.
- Clarify that `typeof` returns the concrete type
- Rename Type Declarations section to Type Annotations,
  avoid confusion with Declaring Types section and distinguish use of "type declaration"
  to mean "declaring new types"
- Removed some of the jargon and wikipedia links to make room for
  a brief alternative `Point{T1,T2}` demonstration.
- Shifted some paragraphs around to reflect their importance,
  and changed the wording in some places.
- Rename type declaration -> annotation in other places (docstrings/comments)

### CODE REVIEW

**Comment by adigitoleo** [comment]
Couldn't find a good place to add a complete custom type family example, and the section is already quite long. Maybe it can go somewhere else eventually.

**Comment by adigitoleo** [comment]
One other thing I left out compared to the  linked issue was talking about DataType, because that has it's own section in Declaring Types.

I do think that Declaring Types needs a bit of work as well, but I'll leave it for another PR.

**Comment by adigitoleo** [comment]
Happy to revert the formatting if it helps. I automatically adopted my markdown habits of breaking at sentence clauses, which I find to produce comfy diffs (except for the first time it's done, like here).

Here's the built html of that page, if it helps: https://x0.at/Bw6y.txt

**Comment by StefanKarpinski** [comment]
I also prefer that style, but it makes the diff here too big and hard to read. You could separate the change into purely formatting and actual content changes as well and then I can just review the content change.

**Comment by adigitoleo** [comment]
Whoops, shouldn't have force-pushed?

**Comment by fingolfin** [comment]
Overall this looks quite good to me! It is a pity that it sat around here for almost 2 years :-(. @adigitoleo sorry for that, if you are still willing to work on it, perhaps you can resolve the conflict?

@StefanKarpinski I think this is relatively nice to review now, perhaps you'd be willing to have another look at it?

File: doc/src/manual/types.md [hunk 1]
```diff
@@ -1,73 +1,59 @@
...
+when the actual values manipulated by the program are available.
+Statically typed languages typically offer faster execution,
+at the cost of type annotations which must be explicitly added by the programmer.
+Dynamic typing, on the other hand, allows for polymorphism,
+which is the ability to write code that can operate on different types.
+Only by explicitly checking types, or when objects fail to support operations at run time,
+are the types of any values ever restricted.
+
+Julia's type system is dynamic, but gains some of the advantages of static type systems
+by allowing optional type annotations. This can be of great assistance
```
Comment (stevengj) [inline doc/src/manual/types.md:17]: Maybe:

> Type annotations are important for defining efficient new data structures, can help enforce correctness and clarity, and can sometimes improve efficiency (although in many contexts the Julia compiler infers types automatically). However, type annotations also play a central role in Julia because they control method dispatch: functions can be extended to new behaviors for different argument types.

File: doc/src/manual/types.md [hunk 2]
```diff
@@ -1,73 +1,59 @@
...
 
-1. As an assertion to help confirm that your program works the way you expect,
-2. To provide extra type information to the compiler, which can then improve performance in some
-   cases
+The default behavior in Julia when type annotations are omitted is to allow values to be of any type.
+When additional expressiveness is needed, however,
+it is easy to gradually introduce explicit annotations into previously "untyped" code.
+Adding annotations serves three primary purposes: to take advantage
+of Julia's powerful multiple-dispatch mechanism, to improve human readability,
+and to catch programmer errors.
```
Comment (stevengj) [inline doc/src/manual/types.md:51]:
```suggestion
Adding annotations serves four primary purpose: to take advantage
of Julia's powerful multiple-dispatch mechanism, to improve human readability,
to catch programmer errors, and to assist the compiler (especially when defining new types).
```
[Nobody expects the Spanish Inquisition!](https://www.youtube.com/watch?v=Cj8n4MfhjUc)

File: doc/src/manual/types.md [hunk 3]
```diff
@@ -78,7 +64,10 @@ julia> (1+2)::Int
 ```
 
 This allows a type assertion to be attached to any expression in-place.
-
+When the type on the right is concrete,
+the value on the left must have that type as its implementation.
+When the type is abstract,
+it suffices for the value to be implemented by a concrete type that is a subtype of the abstract type.
```
Comment (stevengj) [inline doc/src/manual/types.md:70]:
```suggestion
it suffices for the value to have a concrete type that is a subtype of the abstract type.
```

File: doc/src/manual/types.md [hunk 4]
```diff
@@ -595,13 +592,13 @@ julia> struct Point{T}
...
-This declaration defines a new parametric type, `Point{T}`, holding two "coordinates" of type
-`T`. What, one may ask, is `T`? Well, that's precisely the point of parametric types: it can be
-any type at all (or a value of any bits type, actually, although here it's clearly used as a type).
-`Point{Float64}` is a concrete type equivalent to the type defined by replacing `T` in the definition
-of `Point` with [`Float64`](@ref). Thus, this single declaration actually declares an unlimited
-number of types: `Point{Float64}`, `Point{AbstractString}`, `Point{Int64}`, etc. Each of these
-is now a usable concrete type:
+This declaration defines a new parametric type, `Point{T}`,
+holding two "coordinates" of the same type `T`.
+The parameter `T` is a placeholder, which will be replaced by a particular concrete type
```
Comment (stevengj) [inline doc/src/manual/types.md:597]:
```suggestion
The parameter `T` is a placeholder, which will be replaced by a particular type
```
It doesn't have to be concrete, e.g. you can have `Point{Real}`.

**Comment by adigitoleo** [comment]
Thanks for the review. Conflicts should be resolved now. I've addressed the comments, but I'm not sure of the etiquette: should I mark conversations as resolved or leave that to maintainers?

**Comment by oscardssmith** [comment]
I think this is ready to merge, right?

**Comment by adigitoleo** [comment]
Just rebased on lates master, haven't looked at the bulk of the changes in a while but it has been reviewed before. Should be good to go from my side.

File: doc/src/manual/types.md [hunk 5]
```diff
@@ -1,66 +1,59 @@
...
-One particularly distinctive feature of Julia's type system is that concrete types may not subtype
-each other: all concrete types are final and may only have abstract types as their supertypes.
+Types in Julia establish distinctions between different kinds of data,
+and are used by the compiler to infer the intended use of those data.
+Programming languages have traditionally employed one of two quite different type systems:
+static type systems, where expressions must have a computable type before the execution of the program,
+and dynamic type systems, where types are only computed at run time,
+when the actual values manipulated by the program are available.
+Statically typed languages typically offer faster execution,
+at the cost of type annotations which must be explicitly added by the programmer.
```
Comment (stevengj) [inline doc/src/manual/types.md:10]: Explicit annotations are not required in many statically typed languages, e.g. with [Hindley–Milner type systems](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system), which (like Julia) use type inference.

The difference, as @StefanKarpinski has often put it, is that in a static language it is an error if type inference fails, whereas in a dynamic language like Julia the inference is just an optimization.

File: doc/src/manual/types.md [hunk 6]
```diff
@@ -1,66 +1,59 @@
...
+Dynamic typing, on the other hand, allows for polymorphism,
+which is the ability to write code that can operate on different types.
```
Comment (stevengj) [inline doc/src/manual/types.md:12]: Statically typed languages also have polymorphism.

File: doc/src/manual/types.md [hunk 7]
```diff
@@ -567,26 +563,34 @@ See [this FAQ entry](@ref faq-nothing) for more information.
...
+    However, `Foo{Float64} <: Foo{<:Real}` is `true`.
+    This subtle but important point is the reason for the common use of
+    `Container{<:Element}` syntax in type annotations.
 
 ### [Parametric Composite Types](@id man-parametric-composite-types)
 
-Type parameters are introduced immediately after the type name, surrounded by curly braces:
+!!! note
+    For any parametric composite type,
+    both [`isabstracttype`](@ref) and [`isconcretetype`](@ref) will return false.
```
Comment (stevengj) [inline doc/src/manual/types.md:591]:
```suggestion
    both [`isabstracttype`](@ref) and [`isconcretetype`](@ref) will return `false`.
```

File: doc/src/manual/types.md (hunk 5 above)
Comment (adigitoleo) [reply doc/src/manual/types.md:10]: Good points, thanks. How about this for the first paragraph:

> Types in Julia establish distinctions between different kinds of data, and are used by the compiler to infer the intended use of those data. Programming languages have traditionally employed one of two quite different type systems: static type systems, where expressions must have a computable type before the execution of the program, and dynamic type systems, where types are only computed at run time, when the actual values manipulated by the program are available. Statically typed languages typically offer faster execution of programs, while treating any data of unknown or invalid type as an error. In these languages, the programmer is responsible for ensuring the validity of all types of data that the program may encounter. Dynamically typed languages, on the other hand, do not explicitly check types, and types of data only become invalid when they fail to support run-time operations.

File: doc/src/manual/types.md (hunk 6 above)
Comment (adigitoleo) [reply doc/src/manual/types.md:12]: See above suggestion, wherein I removed the misleading reference to polymorphism.

**Comment by adigitoleo** [comment]
I have updated the branch with commits to address the reviews by @stevengj, and the full html folder of the built docs should be downloadable from [this link](https://cryptpad.fr/file/#/2/file/rCHwHbJoYyqxPkTgn-X0x17o/), where the `manual/types.html` file contains the most significant changes.

File: doc/src/manual/types.md (hunk 5 above)
Comment (stevengj) [reply doc/src/manual/types.md]: > Statically typed languages typically offer faster execution of programs

They are not any faster than type-stable code in Julia.   The main argument for static typing, IMO, is safety — by being stricter about types, the compiler can catch more bugs.

But the pros and cons of static vs. dynamic languages is a longstanding debate and a bit of a can of worms here.

Comment (adigitoleo) [reply doc/src/manual/types.md]: I guess by "typically" I probably meant "usually", but I can understand if this unintentionally suggests that Julia is not performant. I'll think about a better wording here.

Comment (fingolfin) [reply doc/src/manual/types.md]: I don't think "usually" is any better

File: doc/src/manual/types.md [hunk 8]
```diff
@@ -78,7 +70,10 @@ julia> (1+2)::Int
 ```
 
 This allows a type assertion to be attached to any expression in-place.
-
+When the type on the right is concrete,
+the value on the left must have that type as its implementation.
```
Comment (akdienes) [inline doc/src/manual/types.md:74]:
```suggestion
```

this is true for dispatch purposes, but not true in general in cases like `x::Int = 'a'`

Comment (adienes) [inline doc/src/manual/types.md:74]:
```suggestion
```

this is true for dispatch purposes, but not true in general in cases like `x::Int = 'a'`

File: doc/src/manual/types.md [hunk 9]
```diff
@@ -1,66 +1,58 @@
...
+by allowing optional type annotations. Type annotations are important for defining efficient new data structures,
+can help enforce correctness and clarity, and can sometimes improve efficiency
+(although in many contexts the Julia compiler infers types automatically.
+Type annotations also play a central role in Julia because they control method dispatch:
+functions can be extended to new behaviours for different argument types.
+Method dispatch is explored in detail in the [Methods](@ref) section, but is rooted in the type system presented here.
+
+All values in a Julia program belong to exactly one concrete type.
+All types belong to a single type tree, i.e. they are all at least related to the [`Any`](@ref) type.
+User-defined types are treated the same as built-in types.
```
Comment (adienes) [inline doc/src/manual/types.md:25]:
```suggestion
```

either this is wrong, or it needs more explanation about what "the same" means here. because there are certainly built-in types which get very special treatment

File: doc/src/manual/types.md [hunk 10]
```diff
@@ -567,26 +562,34 @@ See [this FAQ entry](@ref faq-nothing) for more information.
...
-case. We will discuss them in the following order: first, parametric composite types, then parametric
-abstract types, and finally parametric primitive types.
+each possible combination of parameter values. This is often referred to as "generic programming".
+There are two kinds of explicitly declared parametric types.
+[Parametric composite types](#man-parametric-composite-types)
+declare a family of concrete, [composite types](#Composite-Types),
+whereas [parametric abstract types](#Parametric-Abstract-Types)
+declare a family of [abstract types](#man-abstract-types).
+Other kinds of built-in parametric types such as tuple types are discussed in this section as well.
+Parametric types can also be used in [type annotations](#Type-Annotations).
```
Comment (adienes) [inline doc/src/manual/types.md:572]:
```suggestion
Parametric types can also be used in [type annotations](#Type-Annotations) when the parameters are specified.
```
`x::Foo{T} = 1` is not possible.

File: doc/src/manual/types.md [hunk 11]
```diff
@@ -567,26 +562,34 @@ See [this FAQ entry](@ref faq-nothing) for more information.
...
+
+!!! note
+    Parametric types do not represent a single node in the type graph.
+    Therefore, [`typeof`](@ref) can never return a parametric type name.
+    To dynamically verify a (parametric) subtype relation, use [`isa`](@ref) instead.
+
+!!! warning
+    Type parameters are invariant: even though `Float64 <: Real`,
+    the relation `Foo{Float64} <: Foo{Real}` does NOT hold for any parametric type `Foo`.
```
Comment (adienes) [inline doc/src/manual/types.md:581]:
```suggestion
    the relation `Foo{Float64} <: Foo{Real}` does NOT hold for any parametric type `Foo`.
```

of course except for `Tuple`

File: doc/src/manual/functions.md [hunk 12]
```diff
@@ -106,9 +106,9 @@ the call site that at least one of the arguments (often the first one) is being
     The behavior of a mutating function can be unexpected when a mutated argument shares memory with another argument, a situation known as aliasing (e.g. when one is a view of the other).
     Unless the function docstring explicitly indicates that aliasing produces the expected result, it is the responsibility of the caller to ensure proper behavior on such inputs.
 
-## Argument-type declarations
+## Argument-type annotations
 
-You can declare the types of function arguments by appending `::TypeName` to the argument name, as usual for [Type Declarations](@ref) in Julia.
+You can declare the types of function arguments by appending `::TypeName` to the argument name, as usual for [Type Annotations](@ref) in Julia.
```
Comment (adienes) [inline doc/src/manual/functions.md:111]:
```suggestion
You can annotate the types of function arguments by appending `::TypeName` to the argument name, as usual for [Type Annotations](@ref) in Julia.
```

if the noun is changing `declaration` --> `annotation` then the verb should change `declare` --> `annotate`

File: doc/src/manual/performance-tips.md [hunk 13]
```diff
@@ -250,6 +250,13 @@ better than `IdDict{Type, Vector}`
...
+## Type annotations
+
+In many languages with optional type annotations, adding such annotations is the principal way to
+make code run faster. This is *not* the case in Julia. In Julia, the compiler generally knows
+the types of all function arguments, local variables, and expressions. However, there are a few
+specific instances where annotations are helpful.
+
 ### Avoid fields with abstract type
 
 Types can be declared without specifying the types of their fields:
```
Comment (adienes) [inline doc/src/manual/performance-tips.md:262]:
```suggestion
Types can be annotated without specifying the types of their fields:
```

File: doc/src/manual/types.md [hunk 14]
```diff
@@ -1,66 +1,58 @@
...
+The methods [`isabstracttype`](@ref) and [`isconcretetype`](@ref) can be used
+to check if a type is abstract or concrete.
+They may both return `false` if the type is parametrized but not abstract.
+The [`subtypes`](@ref) and [`supertypes`](@ref) methods
+are also provided for dynamic inspection of the type tree.
+The [`typeof`](@ref) and [`isa`](@ref) methods facilitate checking the type of a value.
 
 Julia's type system is designed to be powerful and expressive, yet clear, intuitive and unobtrusive.
 Many Julia programmers may never feel the need to write code that explicitly uses types. Some
 kinds of programming, however, become clearer, simpler, faster and more robust with declared types.
```
Comment (adienes) [inline doc/src/manual/types.md:43]:
```suggestion
kinds of programming, however, become clearer, simpler, faster and more robust with annotated types.
```

File: doc/src/manual/types.md [hunk 15]
```diff
@@ -78,7 +70,10 @@ julia> (1+2)::Int
 ```
 
 This allows a type assertion to be attached to any expression in-place.
-
+When the type on the right is concrete,
+the value on the left must have that type as its implementation.
+When the type is abstract,
+it suffices for the value to have a concrete type that is a subtype of the abstract type.
 When appended to a variable on the left-hand side of an assignment, or as part of a `local` declaration,
 the `::` operator means something a bit different: it declares the variable to always have the
```
Comment (adienes) [inline doc/src/manual/types.md:78]:
```suggestion
the `::` operator means something a bit different: it annotates the variable to always have the
```

File: doc/src/manual/types.md [hunk 16]
```diff
@@ -595,13 +598,13 @@ julia> struct Point{T}
...
-any type at all (or a value of any bits type, actually, although here it's clearly used as a type).
-`Point{Float64}` is a concrete type equivalent to the type defined by replacing `T` in the definition
-of `Point` with [`Float64`](@ref). Thus, this single declaration actually declares an unlimited
-number of types: `Point{Float64}`, `Point{AbstractString}`, `Point{Int64}`, etc. Each of these
-is now a usable concrete type:
+This declaration defines a new parametric type, `Point{T}`,
+holding two "coordinates" of the same type `T`.
+The parameter `T` is a placeholder, which will be replaced by a particular type
+when an instance of this type is created.
+Thus, this single declaration actually declares an unlimited number of types:
```
Comment (adienes) [inline doc/src/manual/types.md:605]:
```suggestion
Thus, this single annotation actually declares an unlimited number of types:
```

File: doc/src/manual/types.md [hunk 17]
```diff
@@ -1,66 +1,58 @@
...
+Type annotations also play a central role in Julia because they control method dispatch:
+functions can be extended to new behaviours for different argument types.
+Method dispatch is explored in detail in the [Methods](@ref) section, but is rooted in the type system presented here.
+
+All values in a Julia program belong to exactly one concrete type.
+All types belong to a single type tree, i.e. they are all at least related to the [`Any`](@ref) type.
+User-defined types are treated the same as built-in types.
+For additional flexibility, types can be defined as concrete, abstract or parametric.
+Their relation to other types must be explicitly declared.
+One particularly distinctive feature of Julia's type system is that only abstract types can be used as supertypes for a family of related types.
```
Comment (adienes) [inline doc/src/manual/types.md:28]:
```
julia> struct Foo{T} end

julia> Foo{Int} <: Foo
true

julia> isabstracttype(Foo)
false
```

File: doc/src/manual/types.md (hunk 8 above)
Comment (adigitoleo) [reply doc/src/manual/types.md:74]:
```
When the type on the right is concrete,
it must be [`convert`](@ref)able to the type implemented by the value on the left.
```
?

File: doc/src/manual/types.md (hunk 5 above)
Comment (adigitoleo) [reply doc/src/manual/types.md]: > Statically typed languages offer stricter guidance for avoiding unsafe code, such as treating any data of unknown or invalid type as an error. In these languages, the programmer is responsible for ensuring the validity of all types of data that the program may encounter. Writing programs in statically typed languages is, therefore, more laborious, and interactive exploration of data at run-time becomes more difficult. Dynamically typed languages, on the other hand, do not explicitly check types, and types of data only become invalid when they fail to support run-time operations. This provides a more flexible and productive programming experience, at the cost of a greater potential for programmer errors to cause bugs during execution.

?

File: doc/src/manual/types.md (hunk 8 above)
Comment (adigitoleo) [reply doc/src/manual/types.md:74]: This is covered by the rest of the paragraph:

> When appended to a variable on the left-hand side of an assignment, or as part of a `local` declaration, the `::` operator means something a bit different: it declares the variable to always have the specified type, like a type declaration in a statically-typed language such as C. Every value assigned to the variable will be converted to the declared type using [`convert`](@ref):

File: doc/src/manual/types.md (hunk 11 above)
Comment (adigitoleo) [reply doc/src/manual/types.md:581]: I didn't know about that, thanks. Is there a reason why that is the case that we can/should highlight here?

File: doc/src/manual/types.md (hunk 8 above)
Comment (adienes) [reply doc/src/manual/types.md:72]: woops, yes

File: doc/src/manual/types.md (hunk 11 above)
Comment (adienes) [reply doc/src/manual/types.md:579]: I'm not familiar with all the details but my understanding is that the reason for covariance is largely vestigial; it made dispatch easier in the early days of Julia or something


Tuple type covariance is described here in the docs https://docs.julialang.org/en/v1/manual/types/#Tuple-Types-1

and there is an open issue discussing the potential to switch to invariance in a hypothetical breaking release of Julia (although as a gambling man, I wouldn't put my money on that change ever happening) https://github.com/JuliaLang/julia/issues/24614

I would think a quick link to the docs is probably sufficient

**Review by adienes — APPROVED** [review]

**Comment by adigitoleo** [comment]
Hi. Is there still interest in getting this merged? I should be able to take a look in March if more changes are required.

**Comment by adienes** [comment]
personally I think it looks good to go. I'll mark `triage` to see if the next call can take a quick look, not because I think there's anything controversial here but only because rewriting docs on such a fundamental language feature is by nature going to be a sensitive operation.

File: doc/src/manual/performance-tips.md [hunk 18]
```diff
@@ -257,9 +257,16 @@ better than `IdDict{Type, Vector}`
...
+
+In many languages with optional type annotations, adding such annotations is the principal way to
+make code run faster. This is *not* the case in Julia. In Julia, the compiler generally knows
+the types of all function arguments, local variables, and expressions. However, there are a few
+specific instances where annotations are helpful.
+
 ### Avoid fields with abstract type
 
-Types can be declared without specifying the types of their fields:
+Types can be annotated without specifying the types of their fields:
```
Comment (LilithHafner) [inline doc/src/manual/performance-tips.md:269]: Declared is correct here.

File: doc/src/manual/performance-tips.md [hunk 19]
```diff
@@ -257,9 +257,16 @@ better than `IdDict{Type, Vector}`
...
 See also the discussion under [Parametric Types](@ref).
 
+## Type annotations
+
+In many languages with optional type annotations, adding such annotations is the principal way to
+make code run faster. This is *not* the case in Julia. In Julia, the compiler generally knows
+the types of all function arguments, local variables, and expressions. However, there are a few
+specific instances where annotations are helpful.
+
 ### Avoid fields with abstract type
```
Comment (LilithHafner) [inline doc/src/manual/performance-tips.md:267]: This is already present here:
https://github.com/JuliaLang/julia/blob/4a6ada6eb36eaeb9bde9b94c01bcf6f0e3ce7ae6/doc/src/manual/performance-tips.md?plain=1#L209-L214

**Comment by adigitoleo** [comment]
Thanks @LilithHafner, good catch on those and sorry that it took a little while for me to circle back to this.

**Comment by LilithHafner** [comment]
Thanks, we'll discuss this at Triage per @adienes's request!

**Comment by DilumAluthge** [comment]
@LilithHafner Did this ever get discussed at triage?

I see the triage label here, but I don't see a comment summarizing triage's feedback.

**Comment by LilithHafner** [comment]
@DilumAluthge It did not, to my knowledge.

**Comment by DilumAluthge** [comment]
Alright, I've kept the `triage` label, so hopefully it can be discussed during the next triage call.

@LilithHafner Could you summarize what the questions for triage are?

**Comment by StefanKarpinski** [comment]
Triage discussion:

- Don't love the mass rename from "declaration" to "annotation". @xal-0 made the point that we *do* kind of have a distinction between declarations and annotations and the more useful change would be auditing each use and making the distinction intentional and meaningful.
- The major rewrite of the types intro is very hard to review — it mixes real edits with reformatting, which makes the diff hard to read. The rewrite is quite opinionated and not clearly better.
- While there may be improvements to be had, I think that given a choice between the type docs written by the implementers of the type system and a rewrite by someone else (no offense, @adigitoleo), we should favor the writeup from those who designed and implemented it.

We do greatly appreciate the attempt to improve the docs and apologize for the protracted review.
